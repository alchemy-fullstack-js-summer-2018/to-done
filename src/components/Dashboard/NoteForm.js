import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {

  state = {
    editing: false,
    key: null,
    title: '',
    notes: '',
    completed: false
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { key, title, notes, completed } = this.state;
    const note = { title, notes, completed };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ title: '', notes: '', completed: false });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, title, notes, completed } = this.state;
    const { onCancel } = this.props;

    return (
      <div>
        <h3>Add a note:</h3>
        <form onSubmit={this.handleSubmit}>
          <label>Title:<input type="text" name="title" value={title} onChange={this.handleChange}></input></label>
          <label>Notes:<textarea rows="4" cols="50" name="notes" value={notes} onChange={this.handleChange}></textarea></label>
          <label>Completed:<input type="checkbox" name="completed" value={completed} onChange={this.handleChange}></input></label>
          <p>
            <button type="submit">Submit</button>
            {key && <button type="button" onClick={onCancel}>Cancel</button>}
          </p>
        </form>
      </div>
    );
  }
}

export default NoteForm;