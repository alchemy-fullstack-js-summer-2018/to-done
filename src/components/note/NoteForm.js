import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {

  state = {
    editing: false,
    key: null,
    name: '',
    content: ''
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { name, content, key } = this.state;
    const note = { name, content };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ name: '', content: '' });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, name, content } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="name" value={ name } onChange={this.handleChange}/>
        <InputControl name="content" value={ content } onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = (props) => (
  <p>
    <label>
      {props.name}:
      <input {...props} required/>
    </label>
  </p>
);

export default NoteForm;