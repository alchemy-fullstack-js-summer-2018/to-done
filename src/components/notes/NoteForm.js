import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {
  state = {
    editing: false,
    key: null,
    completed: false,
    content: '',
    title: ''
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = event => {
    event.preventDefault();
    const { title, content, completed, key } = this.state;
    const note = { title, content, completed };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ title: '', content: '', completed: false });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, title, content, completed } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="title" value={title} onChange={this.handleChange}/>
        <InputControl name="content" value={content} onChange={this.handleChange}/>
        <label>Completed? </label>
        <input name="completed" value={completed} type="checkbox" onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = props => (
  <p>
    <label>
      {props.name}:
      <input {...props} required />
    </label>
  </p>
);
 
export default NoteForm;