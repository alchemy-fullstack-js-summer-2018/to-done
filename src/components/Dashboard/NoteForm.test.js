import NoteForm from './NoteForm';
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

describe('Note Form', () => {

  it('renders add if no note prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);

    const wrapper = mount(<NoteForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const note = {
      title: 'finish this lab',
      notes: 'you can do it!',
      completed: false
    };

    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: note.title
      }
    });

    wrapper.find('textarea[name="notes"]').simulate('change', {
      target: {
        name: 'notes',
        value: note.notes
      }
    });

    wrapper.find('input[name="completed"]').simulate('change', {
      target: {
        name: 'completed',
        value: note.completed
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(note);

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });

  it('renders correct input on add', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);

    const wrapper = mount(<NoteForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const note = {
      title: 'finish this lab',
      notes: 'you can do it!',
      completed: false
    };

    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'finish this lab'
      }
    });

    wrapper.find('textarea[name="notes"]').simulate('change', {
      target: {
        name: 'notes',
        value: 'you can do it!'
      }
    });

    wrapper.find('input[name="completed"]').simulate('change', {
      target: {
        name: 'completed',
        value: false
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(note);

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });

  it('renders edit if note prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);
    const handleCancel = jest.fn();
    
    const note = {
      key: '4',
      title: 'finish this lab',
      notes: 'you can do it!',
      completed: false
    };

    const wrapper = mount(<NoteForm 
      onComplete={handleComplete}
      onCancel={handleCancel}
      note={note}
    />);

    expect(toJSON(wrapper)).toMatchSnapshot();
    
    wrapper.find('input[name="completed"]').simulate('change', {
      target: {
        name: 'completed',
        value: true
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual({
      ...note,
      completed: true 
    });

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });  
});