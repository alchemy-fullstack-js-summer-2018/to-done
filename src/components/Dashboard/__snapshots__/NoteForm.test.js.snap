// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Form renders add if no note prop 1`] = `
<NoteForm
  onComplete={[MockFunction]}
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value=""
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;

exports[`Note Form renders add if no note prop 2`] = `
<NoteForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "completed": false,
            "notes": "you can do it!",
            "title": "finish this lab",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value="finish this lab"
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value="you can do it!"
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;

exports[`Note Form renders correct input on add 1`] = `
<NoteForm
  onComplete={[MockFunction]}
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value=""
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;

exports[`Note Form renders correct input on add 2`] = `
<NoteForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "completed": false,
            "notes": "you can do it!",
            "title": "finish this lab",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value="finish this lab"
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value="you can do it!"
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;

exports[`Note Form renders edit if note prop 1`] = `
<NoteForm
  note={
    Object {
      "completed": false,
      "key": "4",
      "notes": "you can do it!",
      "title": "finish this lab",
    }
  }
  onCancel={[MockFunction]}
  onComplete={[MockFunction]}
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value="finish this lab"
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value="you can do it!"
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
        <button
          onClick={[MockFunction]}
          type="button"
        >
          Cancel
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;

exports[`Note Form renders edit if note prop 2`] = `
<NoteForm
  note={
    Object {
      "completed": false,
      "key": "4",
      "notes": "you can do it!",
      "title": "finish this lab",
    }
  }
  onCancel={[MockFunction]}
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "completed": true,
            "key": "4",
            "notes": "you can do it!",
            "title": "finish this lab",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <div>
    <h3>
      Add a note:
    </h3>
    <form
      onSubmit={[Function]}
    >
      <label>
        Title:
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label>
        Notes:
        <textarea
          cols="50"
          name="notes"
          onChange={[Function]}
          rows="4"
          value=""
        />
      </label>
      <label>
        Completed:
        <input
          name="completed"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
      </label>
      <p>
        <button
          type="submit"
        >
          Submit
        </button>
        <button
          onClick={[MockFunction]}
          type="button"
        >
          Cancel
        </button>
      </p>
    </form>
  </div>
</NoteForm>
`;
